source/lac/cuda_vector.cu

```diff
@@ -35,51 +35,81 @@ namespace LinearAlgebra
     using ::dealii::CUDAWrappers::block_size;
     using ::dealii::CUDAWrappers::chunk_size;
 
+    namespace
+    {
+      template <typename Number>
+      void
+      delete_device_vector(Number *device_ptr) noexcept
+      {
+        cudaError_t error_code = cudaFree(device_ptr);
+        (void)error_code;
+        AssertNothrow(error_code == cudaSuccess,
+                      dealii::ExcCudaError(cudaGetErrorString(error_code)));
+      }
+
+      template <typename Number>
+      Number *
+      allocate_device_vector(const std::size_t size)
+      {
+        Number *    device_ptr;
+        cudaError_t error_code = cudaMalloc(&device_ptr, size * sizeof(Number));
+        (void)error_code;
+        AssertCuda(error_code);
+        return device_ptr;
+      }
+    } // namespace
+
+
+
     template <typename Number>
     Vector<Number>::Vector()
-      : val(nullptr)
+      : val(nullptr, delete_device_vector<Number>)
       , n_elements(0)
     {}
 
 
 
     template <typename Number>
     Vector<Number>::Vector(const Vector<Number> &V)
-      : n_elements(V.n_elements)
+      : val(allocate_device_vector<Number>(V.n_elements),
+            delete_device_vector<Number>)
+      , n_elements(V.n_elements)
     {
-      // Allocate the memory
-      cudaError_t error_code = cudaMalloc(&val, n_elements * sizeof(Number));
-      AssertCuda(error_code);
       // Copy the values.
-      error_code = cudaMemcpy(val,
-                              V.val,
-                              n_elements * sizeof(Number),
-                              cudaMemcpyDeviceToDevice);
+      cudaError_t error_code = cudaMemcpy(val.get(),
+                                          V.val.get(),
+                                          n_elements * sizeof(Number),
+                                          cudaMemcpyDeviceToDevice);
       AssertCuda(error_code);
     }
 
 
 
     template <typename Number>
-    Vector<Number>::Vector(const size_type n)
-      : val(nullptr)
-      , n_elements(0)
+    Vector<Number> &
+    Vector<Number>::operator=(const Vector<Number> &V)
     {
-      reinit(n, false);
+      if (n_elements < V.n_elements)
+        reinit(V.n_elements);
+
+      n_elements = V.n_elements;
```

Comment:
you could guard this statement with an `else` from the previous `if`. `reinit` will also do the assignment already in the "then" branch